src/rwlock.c

```diff
@@ -2,83 +2,71 @@
 #include "rmalloc.h"
 #include "util/arr_rm_alloc.h"
 #include <assert.h>
+#include "spec.h"
 
-pthread_mutex_t rwLockMutex;
-pthread_rwlock_t RWLock = PTHREAD_RWLOCK_INITIALIZER;
-pthread_key_t _lockKey;
-
-typedef enum lockType { lockType_None, lockType_Read, lockType_Write } lockType;
-
-typedef struct rwlockThreadLocal {
-  size_t locked;
-  lockType type;
-} rwlockThreadLocal;
-
-rwlockThreadLocal** rwlocks;
-
-int RediSearch_LockInit(RedisModuleCtx* ctx) {
-  rwlocks = array_new(rwlockThreadLocal*, 10);
-  pthread_mutex_init(&rwLockMutex, NULL);
-  int err = pthread_key_create(&_lockKey, NULL);
-  if (err) {
-    if (ctx) {
-      RedisModule_Log(ctx, "warning", "could not initialize rwlock thread local");
-    }
-    return REDISMODULE_ERR;
-  }
-  return REDISMODULE_OK;
+void RediSearch_LockInit(struct IndexSpec *sp) {
+  sp->rwLocksData = array_new(rwlockThreadLocal*, 10);
```

Comment:
I am not fully sure why you need this array

Reply:
ohh got it, so you can free all the thread's local data when the index goes down ...